### DISCUSSION TITLE
Difficulties upgrading from v0.2.1 to v0.3.1

### DISCUSSION THREAD

**Discussion by jblindsay** [smartcorelib/smartcore#265] [Q&A] [answered]
Hi again, I recently updated from v0.2.1 and have had to deal with several breaking changes in the API. Most of them were things that I could work around, however, there is one that has caused me much issue and I cannot figure out how to resolve it. `RandomForestClassification` now requires that the target variable satify `Ord`, meaning that floating-point variables are not allowed. I get that, being a classification-type analysis, the target should be nominal/ordinal in nature. However, what I cannot figure out is that `ClassificationMetrics` now has a trait that requires the `RealNumber + FloatNumber` traits. This causes a compilation error for me. What am I not understanding here? Also, out of curiosity, why has the `accuracy()` function in `ClassificationMetrics` been removed? Thanks again.

**Comment by Mec-iS** [comment]
Hi!

As per previous question, please check first the `tests` module in the relative module, in this case [`ensemble/random_forest_classifier.rs`](https://github.com/smartcorelib/smartcore/blob/2d7c05515444ccca7789c37d893d72d1c4a77fc4/src/ensemble/random_forest_classifier.rs#L608). There should be also an example how to use the `accuracy` metrics, it has been moved up into `metrics/accuracy.rs`, not removed.

Consider providing code examples and the compilation error to facilitate debugging when you run into a issue.

Cheers

**Reply by jblindsay** [reply]
Hi @Mec-iS I appreciate you taking the time to respond so quickly. Perhaps I'm being daft here (likely the case!) but I don't see any examples of precision, recall, or f1 in the tests for the random forest classifier. And I cannot seem to use these functions from the `Classificationmetrics` struct because they require the `RealNumber` and `FloatNumber` traits, which the `RandomForestClassification` struct cannot use, because it itself requires the `Ord` trait which f32 and f64 do not implement. I just can't seem to figure out how to rectify this and unfortunately, unless I'm missing it, I don't see those methods used in the tests. I see `accuracy`, and I can get it to work, but not the other metrics. Thanks once again.

<img width="838" alt="Screenshot 2023-05-14 at 9 16 05 AM" src="https://github.com/smartcorelib/smartcore/assets/8441542/2dcceb03-fa0b-4d62-97e1-82d24237cf9c">

**Comment by Mec-iS** [comment]
Hi,
no problem, it is usually better to provide the code snippet and the compiler error as it is much easier to understand the context that trying to explain it in words.

From what I see you are trying to pass a `Vec<usize>` to the first argument of `precision(...)` while the score can only be computed on `f32` or `f64`. In the new version types are a little stricter so the compiler complains. The example in the `precision.rs` module uses this starting data:
```
// predicted data
let y_pred: Vec<f64> = vec![0., 1., 1., 0.];
// ground truth
let y_true: Vec<f64> = vec![0., 0., 1., 1.];
```
As you can see `y_pred` needs now to be a *floating point* value.
You should modify your code to use `f32` as data type, so how you should modify your code depends on how you compute/set the `y_train` value.

**Reply by jblindsay** [reply]
I suppose my issue though is that I can't call:

```
let classifier = RandomForestClassifier::fit(
        &x_train, 
        &y_train, 
        params
    ).expect("Error while developing the model.");
```

With `y_train` as a `Vec<f32>` with the changes in v0.3.1. Because it must implement `Ord` it must be an integer type. Ultimately, with v0.3.1, the classification models must be fit with integer type independent variables, but the metrics require floating-point values.

**Reply by Mec-iS** [reply, marked as answer]
yes the first ones are labels (`usize` that is what Rust uses for indices in arrays) but the second one needs floating point numbers to compute the score.
